kotlinx-coroutines-core/jvm/src/flow/internal/SafeCollector.kt

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2016-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow.internal
+
+import kotlinx.coroutines.flow.*
+import kotlin.coroutines.*
+import kotlin.coroutines.intrinsics.*
+import kotlin.coroutines.jvm.internal.*
+
+@Suppress("UNCHECKED_CAST")
+internal actual class SafeCollector<T> actual constructor(
+    @JvmField internal actual val collector: FlowCollector<T>,
+    @JvmField internal actual val collectContext: CoroutineContext
+) : FlowCollector<T> {
+
+    @JvmField // Note, it is non-capturing lambda, so no extra allocation during init of SafeCollector
+    internal actual val collectContextSize = collectContext.fold(0) { count, _ -> count + 1 }
+    private var lastEmissionContext: CoroutineContext? = null
+    private var intercepted: InterceptedContinuationOwner = InterceptedContinuationOwner()
+    private val emitFun = run {
+        collector::emit as Function2<T, Continuation<Unit>, Any?>
+    }
+
+    /*
+     * Implementor of ContinuationImpl (that will be preserved by ABI implementation nearly forever)
+     * in order to properly control 'intercepted()' lifecycle.
+     */
+    @Suppress("CANNOT_OVERRIDE_INVISIBLE_MEMBER", "INVISIBLE_MEMBER", "INVISIBLE_REFERENCE")
+    private inner class InterceptedContinuationOwner : ContinuationImpl(NoOpContinuation, EmptyCoroutineContext) {
```

Comment:
`SafeCollector` can just extend `ContinuationImpl` directly, without having to allocate a separate `intercepted` field.